### PR TITLE
Blog App : integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Irirwanirira**
## This Pr consist of: 

- [ ] Run tests for user index page:

- I can see the username of all other users.
I-  can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.
- [ ] Run tests for user show page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.
- [ ] Run tests for User post index page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.
- [ ] Run tests for Post show page:

- I can see the post's title.
- I can see who wrote the post.
- I can see how many comments it has.
- I can see how many likes it has.
- I can see the post body.
- I can see the username of each comment.
- I can see the comment each comment left.

- [ ] Before solving  n+1  problem,

before n+1

- [ ] After solving  n+1 problems,

after n+1

### Thank you for reviewing my project

### CODE REVIEW

File: spec/system/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    it 'shows all the properties (name, photo, number of posts) of each user' do
+      visit users_path
+
+      @users.each do |user|
+        expect(page).to have_text(user.name)
+        expect(page).to have_text(user.posts_counter)
+        expect(page).to have_link(href: user_path(user))
+      end
+    end
```
Comment (codepantha) [inline spec/system/user_spec.rb]: - [ ] Please write a test that asserts that you can see the profile picture for each user as specified in the requirements 👇 
image

File: spec/system/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  describe 'Index' do
+    before(:all) do
+      @user = User.first
+      @posts = @user.posts.all
+    end
+
+    it 'I can see the user profile picture' do
+      visit user_posts_path(@user)
+      expect(page).to have_css("img[src*=\"#{@user.photo}\"]")
+    end
```
Comment (codepantha) [inline spec/system/post_spec.rb:17]: - [ ] Please write a test to assert the below requirement 👇 
image

File: spec/system/post_spec.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  describe 'Show' do
+    before(:all) do
+      @user = User.first
+      @post = @user.posts.first
+    end
+
+    it 'I can see the post title' do
+      visit user_post_path(@user, @post)
+      expect(page).to have_content(@post.title)
+    end
```
Comment (codepantha) [inline spec/system/post_spec.rb:81]: - [ ] Please write a test to assert that you can see the comment each commenter left as per requirements 👇 
image

File: spec/system/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      visit user_posts_path(@user)
+      expect(page).to have_content(@user.posts.count)
+    end
+
+    it 'I can see a post title' do
+      visit user_posts_path(@user)
+      @posts.each do |post|
+        expect(page).to have_content(post.title)
+      end
+    end
```
Comment (codepantha) [inline spec/system/post_spec.rb]: - [ ] This particular test is failing as you can see below 👇 Please try to fix it
test fail

File: spec/system/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    it 'I can see the user\'s first 3 posts.' do
+      visit(user_path(@user))
+
+      @user.recent_posts.each do |post|
+        expect(page).to have_text(post.title)
+        expect(page).to have_text(post.text.slice(0, 80))
+        expect(page).to have_link(href: user_post_path(@user.id, post.id))
+      end
+    end
```
Comment (codepantha) [inline spec/system/user_spec.rb]: - [ ] This test is failing as shown in the image below 👇  Please try to make it green ✔️ 
test fail

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi @Irirwanirira 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!




## Highlights 🎯 
✔️ Descriptive PR
✔️ Integration tests 📌
✔️ Good commit history
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/system/user_spec.rb (hunk 5 above)
Comment (Irirwanirira) [reply spec/system/user_spec.rb]: corrected

File: spec/system/post_spec.rb (hunk 4 above)
Comment (Irirwanirira) [reply spec/system/post_spec.rb]: corrected

File: spec/system/post_spec.rb (hunk 2 above)
Comment (Irirwanirira) [reply spec/system/post_spec.rb:17]: done

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi team 👋, 

Well done for implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**